instrumentation/testing/logger.go

```diff
@@ -0,0 +1,184 @@
+package testing
+
+import (
+	"reflect"
+	"runtime"
+	"sync"
+	"testing"
+	"unsafe"
+
+	"github.com/undefinedlabs/go-mpatch"
+
+	"go.undefinedlabs.com/scopeagent/instrumentation"
+)
+
+var (
+	commonPtr          reflect.Type         // *testing.common type
+	patches            []*mpatch.Patch      // patches
+	skippedPointers    = map[uintptr]bool{} // pointers to skip
+	patchPointersMutex sync.Mutex
+	patchPointers      = map[uintptr]bool{} // pointers of patch funcs
+)
+
+func init() {
+	// We get the *testing.common type to use in the patch method
+	var t testing.T
+	typeOfT := reflect.TypeOf(t)
+	if cm, ok := typeOfT.FieldByName("common"); ok {
+		commonPtr = reflect.PtrTo(cm.Type)
+	}
+
+	// We extract all methods pointer of Test, to avoid logging twice
+	var test *Test
+	typeOfTest := reflect.TypeOf(test)
+	for i := 0; i < typeOfTest.NumMethod(); i++ {
+		method := typeOfTest.Method(i)
+		skippedPointers[method.Func.Pointer()] = true
+	}
```

Comment:
Why do you need to do this? Can't you "unpatch" the original function when calling it, and defer its patch, like you do when patching `m.Run`?

Reply:
The use case is with all patched call our `test.Log` instead the normal `t.Log`

`test.Log` will add the events in the span, then call `t.Log`. As it’s patched the events are added again and then the original `t.Log` due the Unpatch.

So the event is added twice. This happens with all our test logging methods.

The solution is two know the pointers of the `test.*` methods and avoid the second call.